services/metadata_service/api/tag.py

```diff
@@ -0,0 +1,139 @@
+from services.data import TaskRow
+from services.data.db_utils import DBResponse
+from services.data.postgres_async_db import AsyncPostgresDB
+from services.metadata_service.api.utils import format_response, \
+    handle_exceptions
+import json
+
+import asyncio
+
+
+class TagApi(object):
+    lock = asyncio.Lock()
+
+    def __init__(self, app):
+        app.router.add_route(
+            "POST",
+            "/tags",
+            self.update_tags,
+        )
+        self._db = AsyncPostgresDB.get_instance()
+
+    def _get_table(self, type):
+        if type == 'flow':
+            return self._db.flow_table_postgres
+        elif type == 'run':
+            return self._db.run_table_postgres
+        elif type == 'step':
+            return self._db.step_table_postgres
+        elif type == 'task':
+            return self._db.task_table_postgres
+        elif type == 'artifact':
+            return self._db.artifact_table_postgres
+        else:
+            raise ValueError("cannot find table for type %s" % type)
+
+    @handle_exceptions
+    @format_response
+    async def update_tags(self, request):
+        """
+        ---
+        description: Update user-tags for objects
+        tags:
+        - Tags
+        parameters:
+        - name: "body"
+          in: "body"
+          description: "body"
+          required: true
+          schema:
+            type: array
+            items:
+              type: object
+              required:
+                - object_type
+                - id
+                - tag
+                - operation
+              properties:
+                object_type:
+                  type: string
+                  enum: [flow, run, step, task, artifact]
+                id:
+                  type: string
+                operation:
+                  type: string
+                  enum: [add, remove]
+                tag:
+                  type: string
+                user:
+                  type: string
+        produces:
+        - application/json
+        responses:
+            "202":
+                description: successful operation. Return newly registered task
+            "404":
+                description: not found
+            "500":
+                description: internal server error
+        """
+        body = await request.json()
+        results = []
+        for o in body:
+            try:
```

Comment:
The multiple updates are not done inside a transaction, so there is a slight peculiarity with the way things might work in case of errors. For example

1. flow/run/step tag updates succeed
2. for whatever reason task tag update fails
3. API responds with an error, even though some records were updated, and some updates were never reached.

Reply:
Yes, this was a replica of how things were happening inside (in our internal version) and was a conscious choice but maybe something worth reconsidering. @ferras may have some more ideas on that.